src/third_party/dartino/two_space_heap.cc

```diff
@@ -368,26 +368,21 @@ bool TwoSpaceHeap::perform_garbage_collection(bool force_compact) {
 
   stack.process(&marking_visitor, old_space(), semi_space);
 
-  bool compact = force_compact || !old_space()->compacting();
+  word regained_by_compacting = old_space()->compute_compaction_destinations();
 
-  if (!compact) {
-    // If the last GC was compacting we don't have fragmentation, so it
-    // is fair to evaluate if we are making progress or just doing
-    // pointless GCs.
-    old_space()->evaluate_pointlessness();
+  if (force_compact || regained_by_compacting <= 0) {
```

Comment:
Negate force_compact -- or even better: Check if you want to compact (force_compact || regained_by_compaction > 0).